docs/sb2d.1
.TH sb2d 1 "7 March 2012" "2.3" "sb2d man page"
.SH NAME
sb2d \- Rule database daemon for Scratchbox2
.SH SYNOPSIS
.B sb2d
[options]
.SH DESCRIPTION
.I sb2d
initializes and manages a rule database inside every Scratchbox2 session.
.PP
Normally
.I sb2d
is started automatically when a session is created by
.I sb2
and will be shut down when the session is
terminated, so there is usually no need to interact with this
daemon directly. However, under some conditions, it might be useful to 
specify options -S, -M or -F for sb2d. That can be done
with the "-x" option of sb2.
.PP
.I sb2d
is responsible for all updates to the database; client processes (i.e.
all processes running inside the session) will use mmap(2) to map the
database image to memory, and will read data from there without
any kind of locking. If a client process needs to add or update data,
it will create a socket and send an RPC message to sb2d, which will
perform the update.
.PP
The database is used to hold several kinds of rules: During session
setup pathmapping rules and exec rules are written to it. Those won't
be modified during session lifetime. But rules related to the virtual
permissions (see sb2(1)) might be added anytime, which might cause
the database file to grow.

.SH OPTIONS

.TP
\-d LEVEL
Enable debug messages.

.TP
\-f
foreground; does not fork (for debugging the daemon).

.TP
\-F MIN_CLIENT_SOCKET_FD
Set minimum client socket filedescriptor value.
Client processes which need to communicate with 
sb2d will create a socket, and then relocate the
file descriptor to a value which is at least
the value defined by this option.
Default is 279.

.TP
\-l FILE
Log messages to FILE. If FILE is "-", writes to stdout.
(note that
.I sb2
redirects stdout of sb2d to $SBOX_SESSION_DIR/sb2d.out
and uses option "-l -")

.TP
\-M MIN_MMAP_ADDR
Set minimum address for the memory mapped database.
Each client process will mmap(2) the database to be above
this address. Default is 0. This can be a 64-bit value,
but note that there can be both 64- and 32-bit client processes,
and the 32-bit processes will drop the most significant
32 bits, and only use the lower part of this value.

.TP
\-n
Don't start server; only initializes the rule database and exits (for debugging the daemon).

.TP
\-p FILE
write process ID to FILE.

.TP
\-s SESSION_DIR
set location of the session directory.

.TP
\-S SIZE
set maximum size of the memory mapped database.
This defines the amount of memory which will be reserved for
the database in
each client process running inside a Scratchbox2 session.
Default is 16 megabytes.

.SH DEBUGGING
A note for developers (of SB2 itself) about debugging:
The rule database file contains binary data. 
There is a debugging tool for printing out the contents,
distributed with the SB2 sources; see utils/sb2-ruletree.c
if you ever need to dig into the low-level details.

.SH FILES
.P
.I $SBOX_SESSION_DIR/*
.P

.SH SEE ALSO
.BR sb2 (1),
.BR sb2-init (1),
.SH BUGS
No known bugs at this time.
.SH AUTHOR
.nf
Lauri T. Aarnio
.fi


debian/changelog
scratchbox2 (2.3.90) unstable; urgency=low

  * No changes to the SW; instead, this version can be considered 
    to be the first release candidate for SB2 version 2.4!
  * Added source of the arch.document (Scratchbox 2 Internals
    and Architecture)

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 05 Sep 2012 14:47:00 +0300

scratchbox2 (2.3.63) unstable; urgency=low

  * Implemented resolution of symbolic links inside
    simulated chroots (absolute symlinks, relative
    symlinks were mostly ok already in 2.3.62)
  * obs-rpm-* modes import /dev and /proc rules
    from the rule library
  * new experimental mapping modes obs-deb-install
    and obs-deb-build; these haven't been tested yet
    and almost certainly will need some tuning.
  * New flag for rules: force_orig_path_unless_chroot
  * Library interface version = 130

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 05 Jun 2012 19:05:00 +0300

scratchbox2 (2.3.62) unstable; urgency=low

  * Initial support for simulating chroot() (ordinary files
    and directories should be OK, but symlinks are not yet
    expected to work correctly in the simulated chroot)
  * Library interface version = 129

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Fri, 01 Jun 2012 21:57:00 +0300

scratchbox2 (2.3.61) unstable; urgency=low

  * A fix to a bug which was introduced in 2.3.60
  * Code cleanups (removed dead code etc. from pathmapping subsystem)

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 30 May 2012 10:37:00 +0300

scratchbox2 (2.3.60) unstable; urgency=low

  * Fixed execution of host binaries which have SUID bit,
    SGID bit or special capabilities: These are now started
    indirectly and SUID/SGID privileges are simulated.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 29 May 2012 11:09:00 +0300

scratchbox2 (2.3.59-1) unstable; urgency=low

  * merge debian packaging from master

 -- Jussi Hakala <jussi.hakala@iki.fi>  Thu, 24 May 2012 23:08:07 +0300

scratchbox2 (2.3.58-1) unstable; urgency=low

  * Bugfix: Fixed handling of target-specific LD_PRELOAD
    and LD_LIBRARY_PATH (those can now be set by e.g.
    "sb2-config setenv LD_LIBRARY_PATH ...")

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 16 May 2012 18:24:00 +0300

scratchbox2 (2.3.57-1) unstable; urgency=low

  * More enhancements to exec functionalty [from Mika Westerberg]
  * Added support for FS rule libraries [Lauri Aarnio]
  * "emulate" mode now imports rules for /dev and /sys
    from the rule library
  * A new rule for /sys/fs/selinux is supposed to hide selinux from
    the target side (in emulate mode), and that way should fix the
    problems caused by a active selinux on the host side.
  * libsb2.so is not linked with -lm anymore

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 08 May 2012 20:16:00 +0300

scratchbox2 (2.3.56-1) unstable; urgency=low

  * Various fixes and enhancements for host-compatible
    targets (i.e. when compiling for intel targets on
    an intel host) [from Mika Westerberg]
  * "emulate" mode: Fixes to /dev mapping rules (e.g.
    "echo >/dev/null" works again with "sb2 -e", and
    not only when "sb2 -eR" is used) [Lauri Aarnio]

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Fri, 27 Apr 2012 18:25:00 +0300

scratchbox2 (2.3.55-1) unstable; urgency=low

  * Added document: Scratchbox 2 Internals and Architecture
    (a .pdf file in the docs subdirectory) 

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 26 Apr 2012 16:41:00 +0300

scratchbox2 (2.3.54-1) unstable; urgency=low

  * A fix to emulate+toolchain mode [from Mika Westerberg]
  * more /dev fixes for the "emulate" mode
  * an additional trick for supporting debootrapping (see
    full comment in the "sb2" script)

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 19 Apr 2012 19:58:00 +0300

scratchbox2 (2.3.53-1) unstable; urgency=low

  * New manual page libsb2_interface(7), generated automatically
    from interface.master
  * fixes for /proc and /dev; "emulate" mode now allows simulated 
    devices to be created under /dev
  * Library interface version = 128
  * Improved checking of rules at startup, refuses to start up if
    there were errors
  * A change to mapping rule format:
    Rules which may not have a valid selector ('dir', 'prefix'
    or 'path' may be nil) need to be marked with 'optional_rule = true'
    (fs_rules: rule_file_interface_version = "105" now). Failing to
    do so will generate warnings to sb2d's log. In the future that
    will be an error which will prevent startup. 
  * + other small fixes.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 18 Apr 2012 17:25:00 +0300

scratchbox2 (2.3.52-1) unstable; urgency=low

  * Virtual permission subsystem: For the simulated root user,
    access(), euidaccess(), eaccess() and faccessat() will
    simulate the checks.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 12 Apr 2012 17:37:00 +0300

scratchbox2 (2.3.51-1) unstable; urgency=low

  * Bugfix to native static binary exec code
  * sb2-show: Removed command "execluafile" and options -x and -X
  * Moved some obsolete C and Lua files to unsupported/obsolete

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 10 Apr 2012 17:57:00 +0300

scratchbox2 (2.3.50-1) unstable; urgency=low

  * No more Lua in libsb2. Lua is now used only in sb2d.
    The runtime "core" of Scratchbox 2 operates now completely
    without the Lua interpreter. Lua is used at session setup
    time from sb2d, but after that, Lua is inactive.
  * All parts of exec postprocessing are now implemented in C,
    except one which was disabled: sbrsh is not anymore available
    as a CPU transparency method.
  * Qemu is the only supported CPU transparency method currently;
    regardless of what has been configured with sb2-init, libsb2
    assumes it is Qemu.
  * The old Lua-related files are still there, and there is 
    dead code in the C files, too. Those are now waiting for a
    bigger cleanup.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 05 Apr 2012 17:51:00 +0300


sb2d/sb2d.c
/*
 * Copyright (C) 2011 Nokia Corporation.
 * Author: Lauri T. Aarnio
 *
 * Licensed under LGPL version 2.1, see top level LICENSE file for details.
*/

/* sb2d, the rule tree server:
 * The one and only process allowed to add data to the rule tree db
 * during an active SB2 session.
 *
 * Other processes may contact this server via a socket (AF_UNIX)
 * and ask the rule tree to be updated; reading from it is allowed
 * without locking, so care must be taken.
 *
 * This implementation assumes that writing 32-bit ints to the
 * memory-mapped file is an atomic operation.
*/

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdarg.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>

#include <sys/socket.h>
#include <sys/un.h>

#include <assert.h>

#include "sb2_server.h"
#include "rule_tree_lua.h"

/* globals */

const char *progname = NULL;
char    *sbox_session_dir = NULL;
char    *pid_file = NULL;


static void write_pid_to_file(pid_t s_pid, const char *pid_file)
{
	if (pid_file) {
		FILE *pidf = NULL;
		pidf = fopen(pid_file, "w");
		if (!pidf) {
			fprintf(stderr, "%s: Failed to open %s for writing\n",
				progname, pid_file);
		} else {
			fprintf(pidf, "%u\n", s_pid);
			fclose(pidf);
		}
	}
}

lua_State *sb2d_lua;

static void load_and_execute_lua_file(const char *filename)
{
	const char *errmsg;

	switch(luaL_loadfile(sb2d_lua, filename)) {
	case LUA_ERRFILE:
                fprintf(stderr, "Error loading %s\n", filename);
                exit(EXIT_FAILURE);
	case LUA_ERRSYNTAX:
                errmsg = lua_tostring(sb2d_lua, -1);
		fprintf(stderr, "Syntax error in %s (%s)\n", filename, 
			(errmsg?errmsg:""));
		exit(EXIT_FAILURE);
	case LUA_ERRMEM:
		fprintf(stderr, "Memory allocation error while "
                        "loading %s\n", filename);
		exit(EXIT_FAILURE);
	case LUA_OK:
		break;
        default:
                fprintf(stderr, "Unknown LUA error "
                        "loading %s\n", filename);
		exit(EXIT_FAILURE);
	}

        switch(lua_pcall(sb2d_lua, 0, 0, 0)) {
        case LUA_ERRRUN:
                fprintf(stderr, "Runtime-Error loading %s (%s)\n",
                        filename, lua_tostring(sb2d_lua, -1));
                exit(EXIT_FAILURE);
	case LUA_ERRMEM:
                fprintf(stderr, "Memory allocation error while "
                        "loading %s (%s) \n",
                        filename, lua_tostring(sb2d_lua, -1));
                exit(EXIT_FAILURE);
        case LUA_OK:
                return;
	}
        exit(1);
}

/* Lua calls this at panic: */
static int sb2_lua_panic(lua_State *l)
{
	fprintf(stderr,
		"sb2d: Lua interpreter PANIC: unprotected error in call to Lua API (%s)\n",
		lua_tostring(l, -1));
#if 0
	sblog_init(); /* make sure the logger has been initialized */
#endif
	SB_LOG(SB_LOGLEVEL_ERROR,
		"Lua interpreter PANIC: unprotected error in call to Lua API (%s)\n",
		lua_tostring(l, -1));
	return 0;
}

static char *read_string_variable_from_lua(const char *name)
{
	char *result = NULL;

	if (name && *name) {
		lua_getglobal(sb2d_lua, name);
		result = (char *)lua_tostring(sb2d_lua, -1);
		if (result) {
			result = strdup(result);
		}
		lua_pop(sb2d_lua, 1);
		SB_LOG(SB_LOGLEVEL_DEBUG,
			"Lua variable %s = '%s', gettop=%d",
			name, (result ? result : "<NULL>"),
			lua_gettop(sb2d_lua));
	}
	return(result);
}

static void initialize_lua(void)
{
	char *main_lua_script = NULL;
	char *lua_if_version = NULL;

	if (asprintf(&main_lua_script, "%s/lua_scripts/init.lua",
	     sbox_session_dir) < 0) {
		SB_LOG(SB_LOGLEVEL_ERROR,
			"%s: asprintf failed to allocate memory", __func__);
		return;
	}
		
	SB_LOG(SB_LOGLEVEL_INFO, "Loading '%s'", main_lua_script);

	sb2d_lua = luaL_newstate();
	lua_atpanic(sb2d_lua, sb2_lua_panic);

	luaL_openlibs(sb2d_lua);
#if 0
	lua_bind_sb_functions(sb2d_lua); /* register our sb_ functions */
#endif
	lua_bind_ruletree_functions(sb2d_lua); /* register our ruletree_ functions */
	lua_bind_sblib_functions(sb2d_lua); /* register our sblib.* functions */

	load_and_execute_lua_file(main_lua_script);

	/* check Lua/C interface version. */
	lua_if_version = read_string_variable_from_lua("sb2d_lua_c_interface_version");
	if (!lua_if_version) {
		SB_LOG(SB_LOGLEVEL_ERROR, "FATAL ERROR: "
			"init.lua script didn't provide"
			" 'sb2d_lua_c_interface_version' identifier!");
		exit(1);
	}
	if (strcmp(lua_if_version, SB2D_LUA_C_INTERFACE_VERSION)) {
		SB_LOG(SB_LOGLEVEL_ERROR, "FATAL ERROR: "
			"init.lua script interface version mismatch:"
			" script has '%s', but '%s' was expected",
			lua_if_version, SB2D_LUA_C_INTERFACE_VERSION);
		exit(1);
	}
	free(lua_if_version);

	SB_LOG(SB_LOGLEVEL_INFO, "lua initialized.");
	SB_LOG(SB_LOGLEVEL_NOISE, "gettop=%d", lua_gettop(sb2d_lua));

	free(main_lua_script);
}

char *execute_init2_script(void)
{
	char	*init2_script = NULL;
	char	*result = NULL;

	if (asprintf(&init2_script, "%s/lua_scripts/init2.lua",
	     sbox_session_dir) < 0) {
		SB_LOG(SB_LOGLEVEL_ERROR,
			"%s: asprintf failed to allocate memory", __func__);
		return(NULL);
	}
		
	SB_LOG(SB_LOGLEVEL_INFO, "Loading '%s'", init2_script);

	load_and_execute_lua_file(init2_script);

	/* get result. */
	result = read_string_variable_from_lua("init2_result");
	if (!result) {
		SB_LOG(SB_LOGLEVEL_ERROR, "init2 scripts didn't provide"
			" 'init2_result'!");
	}
	return(result);
}

static long long parse_num(const char *cp)
{
	long	l;
	char	*end = NULL;

	if (!cp || !*cp) {
		fprintf(stderr, "ERROR: A numeric argument is required");
		exit(1);
	}
	l = strtoll(cp, &end, 0);
	if (!end || (end == cp) || (*end != '\0')) {
		fprintf(stderr, "ERROR: %s is not a valid numeric argument", cp);
		exit(1);
	}
	return(l);
}

int main(int argc, char *argv[])
{
	int	opt;
	int	start_server = 1;
	int	backgroud_server = 1;
	char	*debug_level = NULL;
	char	*debug_file = NULL;
	char	*rule_tree_path = NULL;
	uint32_t max_size = 16*1024*1024; /* default 16MB */
	uint64_t min_mmap_addr = 0;
	int	min_client_socket_fd = 279;

	progname = argv[0];

	/* check data types: the rule tree DB uses plain types for
	 * certain data types; these assertions guarantee
	 * that SB2 startup fails if the OS is changed someday
	 * in the future, and data would not fit anumore to the DB...
	*/
	assert(sizeof(uint64_t) >= sizeof(dev_t));
	assert(sizeof(uint64_t) >= sizeof(ino_t));
	assert(sizeof(uint32_t) >= sizeof(uid_t));
	assert(sizeof(uint32_t) >= sizeof(gid_t));
	assert(sizeof(uint32_t) >= sizeof(mode_t));

	while ((opt = getopt(argc, argv, "d:l:s:p:nfS:M:F:")) != -1) {
		switch (opt) {
		case 'd':
			debug_level = strdup(optarg);
			break;
		case 'l':
			debug_file = strdup(optarg);
			break;
		case 's':
			sbox_session_dir = strdup(optarg);
			break;
		case 'p':
			pid_file = strdup(optarg);
			break;
		case 'n': /* don't start server */
			start_server = 0;
			break;
		case 'f': /* foreground server (don't fork) */
			backgroud_server = 0;
			break;
		case 'S':
			max_size = parse_num(optarg);
			break;
		case 'M':
			min_mmap_addr = parse_num(optarg);
			break;
		case 'F':
			min_client_socket_fd = parse_num(optarg);
			break;
		default:
			fprintf(stderr, "Illegal option\n");
			exit(1);
		}
	}
	/* init logging before doing anything else,
	 * otherwise the logger will auto-initialize itself.
	 * if debug_level and/or debug_file is NULL, logger
	 * will read the values from env.vars. */
	sblog_init_level_logfile_format(debug_level, debug_file, NULL);

	if (!sbox_session_dir) {
		fprintf(stderr, "ERROR: "
			"Option '-s session_dir' is mandatory.\n");
		exit(1);
	}

	/* ----- Initialization ----- */

	SB_LOG(SB_LOGLEVEL_DEBUG, "Initializing");

	if (asprintf(&rule_tree_path, "%s/RuleTree.bin", sbox_session_dir) < 0) {
		exit(1);
	}

	if (create_ruletree_file(rule_tree_path,
		max_size, min_mmap_addr, min_client_socket_fd) < 0) {

		SB_LOG(SB_LOGLEVEL_ERROR, "Failed to create rule tree file (%s)",
			rule_tree_path);
		fprintf(stderr, "Failed to create rule DB!\n");
		exit(1);
	}
	SB_LOG(SB_LOGLEVEL_DEBUG, "Rule tree file opened & mapped to memory");

	initialize_lua();

	/* ----- Server ----- */
	if (start_server) {
		pid_t worker_pid;

		SB_LOG(SB_LOGLEVEL_DEBUG, "Initializing server");
		create_server_socket();

		/* Write PID to a file */
		if (backgroud_server) {
			if ((worker_pid = fork()) != 0) {
				/* parent */
				SB_LOG(SB_LOGLEVEL_DEBUG, "Parent process");
				write_pid_to_file(worker_pid, pid_file);
				return(0);
			}
		} else {
			write_pid_to_file(getpid(), pid_file);
		}
		
		/* enter the server loop. 
		 * ruletree_server() returns when the socket has been
		 * deleted and it is time to shut down. */
		ruletree_server();
	}
	return(0);
}

